### PR TITLE
Added Travis configuration for building Docker images on for the Hub

### PR DESCRIPTION
**Pull request by moritzheiber**
This PR implements building Docker images for either `master` or release tags on TravisCI to work around the issue of the Docker Hub using an outdated version of Docker.

Before merging this PR @Gargron needs to run the following command inside project directory:

```sh
$ travis encrypt DOCKER_HUB_PASSWORD=<their-docker-hub-password> --add
```
and check-in the resulting `.travis.yml` for this PR, otherwise the pushes to the Hub will fail (since they require a valid login).

### CODE REVIEW

**Comment by nightpool** [comment]
shouldnt gargron run that command and push the commit to this PR BEFORE
merging it?

**Comment by moritzheiber** [comment]
@nightpool True, I completely forgot about commit contributions on PRs. It's been a long day :disappointed: 
@Gargron You can do this before merging the PR, obviously, as @nightpool suggested.

File: .travis.yml
Comment (ineffyble) [inline .travis.yml]: Can you clarify what this regex is?

**Review by ineffyble — APPROVED** [review]
LGTM

File: .travis.yml
Comment (moritzheiber) [reply .travis.yml]: It matches branches (yes, for some reason Travis detects tags as branches) with release criteria, i.e. "v2.3.0rc1", or "v.2.3.4-test".

File: .travis.yml
Comment (akihikodaki) [inline .travis.yml]: Redundant quotes.

File: deploy_hub.sh
Comment (akihikodaki) [inline deploy_hub.sh]: `latest` is confusing. It sounds like the latest "release", but it is actually not release. I suggest just to use `TRAVIS_BRANCH`; a user should know that something from Mastodon named `master` is unstable.

File: deploy_hub.sh
Comment (akihikodaki) [inline deploy_hub.sh]: I do not think the last line is necessary.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

**Comment by akihikodaki** [comment]
I wonder how long it takes since it builds `libiconv`.

File: .travis.yml
Comment (moritzheiber) [reply .travis.yml]: Right, well spotted. It worked without having removed it, but this can be considered obsolete. I'll fix it :+1:

File: .travis.yml
Comment (moritzheiber) [reply .travis.yml]: This went through Travis' own linter (which is why some parts of the file have changed) and I'd rather like to leave it that way.

File: deploy_hub.sh
Comment (moritzheiber) [reply deploy_hub.sh]: I'm just trying to replicate current behavior. Currently `latest` points at the tip of master (although it hasn't been built in a while due to the bug this PR is trying to solve), as far as I can tell, hence the same behavior in here.

File: deploy_hub.sh
Comment (moritzheiber) [reply deploy_hub.sh]: It is if `TRAVIS_TAG` is actually set. You would want a tagged release to become `latest` as well. Also, this is a no-op in the daemon if it detects that the digest has already been pushed.

**Comment by moritzheiber** [comment]
@akihikodaki Are you okay with the explanation I've given to your requests above? Also, the overhead for compiling libiconv is roughly 30 seconds. Seeing as the majority of our Travis runs are spend on installing system dependencies and running RSpec tests right now I wouldn't consider it a large overhead.

(Also, this isn't done for PRs, just for `master` and release tags)

File: deploy_hub.sh
Comment (ineffyble) [reply deploy_hub.sh]: `latest` has a tradition of being bleeding edge in any case - https://medium.com/@mccode/the-misunderstood-docker-tag-latest-af3babfd6375

Comment (akihikodaki) [reply deploy_hub.sh]: > I'm just trying to replicate current behavior. Currently `latest` points at the tip of master (although it hasn't been built in a while due to the bug this PR is trying to solve), as far as I can tell, hence the same behavior in here.

I do not think there is any strict rule yet. It is a good chance to consider about it well.

> If latest should be bleeding edge (I'd be in favor of it) or not isn't for this PR to decide.

It is if there is no rule yet.

> `latest` has a tradition of being bleeding edge in any case

Bleeding edge "release", maybe? Indeed many Docker images are just for releases and it is hard to find examples of CI.

**Review by akihikodaki — APPROVED** [review]

**Comment by akihikodaki** [comment]
I have concerns about `latest` but I think it is not that important to block merging this.

**Comment by akihikodaki** [comment]
Well, is an explanation why it is rejected provided? I'm just curious.

**Comment by moritzheiber** [comment]
@akihikodaki Hey, sorry! I told Eugen to close it because it was essentially superseded by #6662. Using the "native" Docker Hub environment, along with what Eugen has now declared to be `edge` and `latest`, makes sense, and we don't need to rely on Travis for building our images.

I hope that clears it up. I should've posted a follow-up earlier.

**Comment by akihikodaki** [comment]
@moritzheiber Thank you for your reply. I understood now.

**Pull request closed without merging**
